src/bun.js/child_process.exports.js

```diff
@@ -1,3 +1,1622 @@
+      stdout = ArrayPrototypeJoin.call(_stdout, "");
+    } else {
+      stdout = BufferConcat(_stdout);
+    }
+    if (encoding || (child.stderr && child.stderr.readableEncoding)) {
+      stderr = ArrayPrototypeJoin.call(_stderr, "");
+    } else {
+      stderr = BufferConcat(_stderr);
+    }
+
+    // TODO: Make this check code === 0 when Bun.spawn fixes exit code issue
+    if (!ex && code >= 0 && signal === null) {
+      callback(null, stdout, stderr);
+      return;
+    }
+
+    if (args?.length) cmd += ` ${ArrayPrototypeJoin.call(args, " ")}`;
+
+    if (!ex) {
+      ex = genericNodeError(`Command failed: ${cmd}\n${stderr}`, {
+        // code: code < 0 ? getSystemErrorName(code) : code, // TODO: Add getSystemErrorName
+        code: code,
+        killed: child.killed || killed,
+        signal: signal,
+      });
+    }
+
+    ex.cmd = cmd;
+    callback(ex, stdout, stderr);
+  }
+
+  function errorHandler(e) {
+    ex = e;
+
+    if (child.stdout) child.stdout.destroy();
+    if (child.stderr) child.stderr.destroy();
+
+    exitHandler();
+  }
+
+  function kill() {
+    if (child.stdout) child.stdout.destroy();
+    if (child.stderr) child.stderr.destroy();
+
+    killed = true;
+    try {
+      child.kill(options.killSignal);
+    } catch (e) {
+      ex = e;
+      exitHandler();
+    }
+  }
+
+  if (options.timeout > 0) {
+    timeoutId = setTimeout(function delayedKill() {
+      kill();
+      timeoutId = null;
+    }, options.timeout);
+  }
+
+  if (child.stdout) {
+    if (encoding) child.stdout.setEncoding(encoding);
+
+    child.stdout.on("data", function onChildStdout(chunk) {
+      // Do not need to count the length
+      if (options.maxBuffer === Infinity) {
+        ArrayPrototypePush.call(_stdout, chunk);
+        return;
+      }
+      const encoding = child.stdout.readableEncoding;
+      const length = encoding
+        ? Buffer.byteLength(chunk, encoding)
+        : chunk.length;
+      const slice = encoding
+        ? (buf, ...args) => StringPrototypeSlice.call(buf, ...args)
+        : (buf, ...args) => buf.slice(...args);
+      stdoutLen += length;
+
+      if (stdoutLen > options.maxBuffer) {
+        const truncatedLen = options.maxBuffer - (stdoutLen - length);
+        ArrayPrototypePush.call(_stdout, slice(chunk, 0, truncatedLen));
+
+        ex = new ERR_CHILD_PROCESS_STDIO_MAXBUFFER("stdout");
+        kill();
+      } else {
+        ArrayPrototypePush.call(_stdout, chunk);
+      }
+    });
+  }
+
+  if (child.stderr) {
+    if (encoding) child.stderr.setEncoding(encoding);
+
+    child.stderr.on("data", function onChildStderr(chunk) {
+      // Do not need to count the length
+      if (options.maxBuffer === Infinity) {
+        ArrayPrototypePush.call(_stderr, chunk);
+        return;
+      }
+      const encoding = child.stderr.readableEncoding;
+      const length = encoding
+        ? Buffer.byteLength(chunk, encoding)
```

Comment:
`Buffer.byteLength` can be very expensive because it gets the exact size which means an extra pass over the string

It would be faster in most cases to instead say something like "is the current length * 4 more than the maximum buffer? then lets get the precise byteLength"